### PR TITLE
Allow Azure client to set linked work item

### PR DESCRIPTION
**Pull request by JamieMagee**
This PR adds the ability to link work items when create a pull request in Azure DevOps.

The API is pretty forgiving, and allows the JSON `"workItemRefs": [{"id": null}]`

API docs: https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20requests/create?view=azure-devops-rest-5.0

### CODE REVIEW

**Comment by JamieMagee** [comment]
Hi @jurre, another PR for you to review :smile:

File: .rubocop.yml [hunk 1]
```diff
@@ -101,6 +101,7 @@ Metrics/BlockLength:
     - '*/dependabot-*.gemspec'
 
 Metrics/ParameterLists:
+  Max: 6
```
Comment (jurre) [inline .rubocop.yml:104]: I'm not sure that we want to globally increase this, could we add inline exceptions for the methods that need it? Globally relaxing such rules is a bit of a slippery slope IMO, and we're already not in fantastic shape regarding linting in this project.

File: common/lib/dependabot/pull_request_creator.rb [hunk 2]
```diff
@@ -33,7 +33,8 @@ def initialize(source:, base_commit:, dependencies:, files:, credentials:,
                    branch_name_separator: "/", branch_name_prefix: "dependabot",
                    label_language: false, automerge_candidate: false,
                    github_redirection_service: "github-redirect.dependabot.com",
-                   custom_headers: nil, require_up_to_date_base: false)
+                   custom_headers: nil, require_up_to_date_base: false,
+                   work_item: nil)
```
Comment (jurre) [inline common/lib/dependabot/pull_request_creator.rb:37]: I wonder if it makes sense to start splitting up these PR creators based on the provider, there's more and more conditional logic / arguments, increasing the complexity and making it harder to add features and test things.

Not something that I'd expect in this PR, but more food for thought for ourselves cc @brrygrdn @feelepxyz

Comment (feelepxyz) [reply common/lib/dependabot/pull_request_creator.rb:37]: Ah yeah would be ace to think about splitting these out. `work_item` here is slightly confusing as it's only for azure. Or could we add a generic `provider_metadata` hash that we pass to all pull request creators?

Comment (JamieMagee) [reply common/lib/dependabot/pull_request_creator.rb:37]: @feelepxyz Totally agree with the idea. However, is this something you would like to see as part of this PR, or as a separate PR? It feels like a bit of a larger refactoring task in my opinion.

File: .rubocop.yml (hunk 1 above)
Comment (JamieMagee) [reply .rubocop.yml:104]: I do agree. I think this could be solved with the `provider_metadata` suggestion from @feelepxyz below.

File: common/lib/dependabot/pull_request_creator.rb (hunk 2 above)
Comment (feelepxyz) [reply common/lib/dependabot/pull_request_creator.rb:37]: @JamieMagee yeah lets leave a larger refactor out of this PR but thoughts on adding a new `provider_metadata` hash that gets passed in the Azure PR creator and then extract the `work_item` key?

Comment (JamieMagee) [reply common/lib/dependabot/pull_request_creator.rb:37]: @feelepxyz I pushed an update that adds a `provider_metadata` hash currently just targeting the Azure specific properties, but this can easily be expanded to the other providers.

I think I can iterate on this a bit more. I think I'll pass `provider_metadata` directly through to the Azure client.

Comment (jurre) [reply common/lib/dependabot/pull_request_creator.rb:37]: Maybe I misunderstood, but what I was thinking, was that it would be better to to pass the Azure specific `work_item` to `Dependabot::PullRequestCreator`, but passing it to the Azure client explicitly is fine.

So my thinking was that we change `Dependabot::PullRequestCreator#initialize` to:

```ruby
    def initialize(source:, base_commit:, dependencies:, files:, credentials:,
                   branch_name_separator: "/", branch_name_prefix: "dependabot",
                   label_language: false, automerge_candidate: false,
                   github_redirection_service: "github-redirect.dependabot.com",
                   custom_headers: nil, require_up_to_date_base: false)
                   custom_headers: nil, require_up_to_date_base: false,
                   provider_metadata: {})
```

We'd then pass the `work_item` in that metadata:

```ruby
Dependabot::PullRequestCreator.new(
# ...
provider_metadata: { work_item: "my-work-item" }
)
```

And then when creating the Azure client we pull it out:

```ruby
    Azure.new(
        source: source,
        branch_name: branch_namer.new_branch_name,
        base_commit: base_commit,
        credentials: credentials,
        files: files,
        commit_message: message_builder.commit_message,
        pr_description: message_builder.pr_message,
        pr_name: message_builder.pr_name,
        author_details: author_details,
        labeler: labeler,
        work_item: provider_metadata[:work_item]
      )
```

That way we don't pollute the public interface of the generic PR creator with implementation specific arguments. @feelepxyz is that also what you were thinking?

Comment (JamieMagee) [reply common/lib/dependabot/pull_request_creator.rb:37]: Ah, okay. I thought you want to just pass in an object and extract properties as needed 🤦. Just to clarify you want the public interface to have all common properties, and then a has for platform specific properties?

File: common/lib/dependabot/pull_request_creator.rb [hunk 3]
```diff
@@ -142,7 +144,8 @@ def azure_creator
         pr_description: message_builder.pr_message,
         pr_name: message_builder.pr_name,
         author_details: author_details,
-        labeler: labeler
+        labeler: labeler,
+        work_item: provider_metadata&.fetch(:work_item)
```
Comment (feelepxyz) [inline common/lib/dependabot/pull_request_creator.rb:148]: Think this should default to nil otherwise it will raise

```suggestion
        work_item: provider_metadata&.fetch(:work_item, nil)
```

File: common/lib/dependabot/pull_request_creator/azure.rb [hunk 4]
```diff
@@ -46,9 +47,7 @@ def azure_client_for_source
       end
 
       def branch_exists?
-        @branch_ref ||= azure_client_for_source.branch(branch_name)
-
-        @branch_ref
```
Comment (feelepxyz) [inline common/lib/dependabot/pull_request_creator/azure.rb:51]: Is this change intentional?

File: common/spec/dependabot/pull_request_creator_spec.rb [hunk 5]
```diff
@@ -45,6 +46,7 @@
   let(:source) { Dependabot::Source.new(provider: "github", repo: "gc/bump") }
   let(:files) { [gemfile] }
   let(:base_commit) { "basecommitsha" }
+  let(:provider_metadata) { { work_item: 123 } }
```
Comment (feelepxyz) [inline common/spec/dependabot/pull_request_creator_spec.rb:49]: Maybe better to set this to nil here and only set it with a `work_item` in the `with an Azure source` context
```suggestion
  let(:provider_metadata) { nil }
```

File: common/lib/dependabot/pull_request_creator/azure.rb (hunk 4 above)
Comment (JamieMagee) [reply common/lib/dependabot/pull_request_creator/azure.rb:51]: Yes, RubyMine highlighted this as an unnecessary store before return. It's now the same implementation as the GitHub client, however the GitLab and Code Commit clients still use this pattern.

Comment (feelepxyz) [reply common/lib/dependabot/pull_request_creator/azure.rb:51]: ah yes it's only used once from `create` so seems good to me 👍

File: common/lib/dependabot/clients/azure.rb [hunk 6]
```diff
@@ -153,8 +153,9 @@ def create_commit(branch_name, base_commit, commit_message, files,
           "/pushes?api-version=5.0", content.to_json)
       end
 
+      # rubocop:disable Metrics/ParameterLists
       def create_pull_request(pr_name, source_branch, target_branch,
-                              pr_description, labels)
+                              pr_description, labels, work_item)
```
Comment (feelepxyz) [inline common/lib/dependabot/clients/azure.rb:158]: Somewhat annoying but just thinking about existing users who might be calling these methods directly so should probably make these optional so we don't introduce breaking changes

```suggestion
                              pr_description, labels, work_item = nil)
```

File: common/lib/dependabot/pull_request_creator/azure.rb [hunk 7]
```diff
@@ -8,11 +8,11 @@ class PullRequestCreator
     class Azure
       attr_reader :source, :branch_name, :base_commit, :credentials,
                   :files, :commit_message, :pr_description, :pr_name,
-                  :author_details, :labeler
+                  :author_details, :labeler, :work_item
 
       def initialize(source:, branch_name:, base_commit:, credentials:,
                      files:, commit_message:, pr_description:, pr_name:,
-                     author_details:, labeler:)
+                     author_details:, labeler:, work_item:)
```
Comment (feelepxyz) [inline common/lib/dependabot/pull_request_creator/azure.rb:15]:
```suggestion
                     author_details:, labeler:, work_item: nil)
```

**Review by feelepxyz — COMMENTED** [review]
Looks good to me! Added some final comments about making the arguments optional so we don't accidentally introduce any breaking changes to existing users of the Azure PR creator.

**Review by jurre — APPROVED** [review]
